### PR TITLE
Generalize multipart upload implementations in S3 client

### PR DESCRIPTION
**Pull request by xemul**
There are two currently -- upload_sink_base and do_upload_file. This PR merges as much code as possible (spoiler: it's already mostly copy-n-pase-d, so squashing is pretty straightforward)

### CODE REVIEW

File: utils/s3/client.cc [hunk 1]
```diff
@@ -500,14 +500,24 @@ class client::upload_sink_base : public data_sink_impl {
...
         , _object_name(std::move(object_name))
         , _tag(std::move(tag))
     {
     }
 
+public:
+    unsigned parts_count() const noexcept { return _part_etags.size(); }
+};
+
+class client::upload_sink_base : public multipart_upload, public data_sink_impl {
```
Comment (tchaikov) [inline utils/s3/client.cc:514]: if `upload_sink_base` is but *using* `multipart_upload`, can we private-inherit from it? the same applies to `do_upload_file`.

**Review by tchaikov — COMMENTED** [review]
modulo one nit, lgtm. and thanks, less code == less trouble.

File: utils/s3/client.cc (hunk 1 above)
Comment (xemul) [reply utils/s3/client.cc:514]: No, it complains in a tricky way. I made upload_sink_base protected inherit from multipart_upload

```
utils/s3/client.cc:786:72: error: '_object_name' is a protected member of 's3::client::multipart_upload'
  786 |     s3l.trace("PUT part {} from {} (upload id {})", part_number, piece._object_name, _upload_id);
      |                                                                        ^
```

that's the code

```c++
future<> client::multipart_upload::upload_part(std::unique_ptr<upload_sink> piece_ptr) {
    if (!upload_started()) {
        co_await start_upload();
    }

    auto& piece = *piece_ptr;
    unsigned part_number = _part_etags.size();
    _part_etags.emplace_back();
    s3l.trace("PUT part {} from {} (upload id {})", part_number, piece._object_name, _upload_id);
```

the multipart_upload tries to access field from one of its derived classes and fails :(

Comment (xemul) [reply utils/s3/client.cc:514]: But private inheriting of do_upload_file works fine

**Pull request closed without merging**